compose/src/main/java/com/stripe/android/compose/elements/ElementState.kt

```diff
@@ -0,0 +1,39 @@
+package com.stripe.android.compose.elements
+
+/**
+ * This represents the different states a field can be in,
+ * in each of these cases there might be a reason to show the
+ * error in a different way
+ */
+abstract class ElementState {
+
+    /**
+     * This indicate the Element state has reached the max number of characters
+     * and focus should shift forward.
+     */
+    abstract fun isFull(): Boolean
+
+    /**
+     * Indicates element is valid and field extraction can happen
+     * and be used to create PaymentMethod Parameters
+     */
+    abstract fun isValid(): Boolean
+
+    /**
+     * If in a state where isValid() returns false it indicates the element cannot be
+     * used for extraction.  This function should return the error message.
+     */
+    abstract fun getErrorMessageResId(): Int?
+
+    open class ElementStateValid : ElementState() {
+        override fun isValid(): Boolean = true
+        override fun isFull(): Boolean = false
+        override fun getErrorMessageResId(): Int? = null
+    }
+
+    open class ElementStateError(private val stringResId: Int) : ElementState() {
+        override fun isValid(): Boolean = false
+        override fun isFull(): Boolean = false
+        override fun getErrorMessageResId(): Int? = stringResId
+    }
```

Comment:
There are quite a few overlaps between the errors in the fields (e.g. `BlankAndRequired`). Would it make sense to keep a single instance of these errors and share it between all the different fields? I know that each field might have its own way of displaying an error so I don't know how viable this approach would be with respect to custom error types.

Reply:
I spent a fair amount of time weighing the decisions here, and actually started with one element state for all the fields, but I didn't like it because some of those states aren't relevant.  In terms of where change is likely to occur I think it is in what states the fields can be in, and how they want to handle those states.  By keeping the states contained in the config we enable those changes.

Reply:
Just to add some examples to this:
Credit card number has malformed where you enter a '0' as the first number, it should immediately show it as an error (not only when focus is lost).
Field that have a limited number of characters allowed (such as CVC and card number), it might make sense to still allow the user to add extra characters even though we think we reached the limited number of characters.  This will make it so that if we have the validation wrong we aren't preventing payment.  (Final validation would still happen on the server).

These are a couple of examples, but they have made me believe that the states and how errors are displayed is pretty field dependent.

One idea I considered was having a default shouldShowError method that has the default behavior, and allow each field to override it with it's own states and behaviors.